src/SpendPermissionManager.sol

```diff
@@ -279,15 +302,44 @@ contract SpendPermissionManager is EIP712 {
         _transferFrom(spendPermission.token, spendPermission.account, spendPermission.spender, value);
     }
 
+    /// @notice Revokes a spend permission and approves a new one atomically.
+    ///
+    /// @dev Enforces that the last updated period of the permission being revoked matches the last valid updated period
+    ///      submitted as an argument. This is to prevent frontrunning `replace` with additional last-minute spends.
+    ///
+    /// @dev The `account` of the permissions must match, but the remaining fields can differ.
+    ///
+    /// @dev Can only be called by the `account` of a permission.
+    ///
+    /// @param permissionToRevoke Details of the spend permission to revoke.
+    /// @param permissionToApprove Details of the spend permission to approve.
+    /// @param lastValidUpdatedPeriod Last valid updated period for the spend permission being revoked.
+    function replace(
+        SpendPermission calldata permissionToRevoke,
+        SpendPermission calldata permissionToApprove,
+        PeriodSpend calldata lastValidUpdatedPeriod
+    ) external requireSender(permissionToApprove.account) requireSameAccount(permissionToRevoke, permissionToApprove) {
+        // validate that no spending has occurred since the last updated period passed to the function
+        PeriodSpend memory lastUpdatedPeriod = getLastUpdatedPeriod(permissionToRevoke);
+        if (
+            lastUpdatedPeriod.spend != lastValidUpdatedPeriod.spend
+                || lastUpdatedPeriod.start != lastValidUpdatedPeriod.start
+                || lastUpdatedPeriod.end != lastValidUpdatedPeriod.end
+        ) {
+            revert InvalidLastUpdatedPeriod(lastUpdatedPeriod, lastValidUpdatedPeriod);
+        }
+        // revoke old and approve new spend permissions
+        _revoke(permissionToRevoke);
+        _approve(permissionToApprove);
```

Comment:
do we want to enforce that these two are not the same permission?

Reply:
personally I don't... we don't enforce that batches don't contain duplicates, nor do we enforce that permissions being approved aren't already approved (or permissions being revoked aren't already revoked), and in general have taken the approach of avoiding validation unless there's a risk of inconsistent state onchain. adding validation here would feel arbitrary.